src/math_parser_diag.cpp

```diff
@@ -443,6 +523,190 @@ std::function<void( dialogue &, double )> spell_exp_ass( char scope,
+                    return it->second;
+                } else {
+                    return 0.0;
+                }
+            } else if( !spell_str.empty() ) {
+                std::map<spell_id, double>::iterator it = ch->magic->caster_level_adjustment_by_spell.find(
+                            spell_id( spell_str ) );
+                if( it != ch->magic->caster_level_adjustment_by_spell.end() ) {
+                    return it->second;
+                } else {
+                    return 0.0;
+                }
+            } else {
+                return ch->magic->caster_level_adjustment;
+            }
+        }
+        return 0.0;
+    };
+}
+
+std::function<void( dialogue &, double )> spell_level_adjustment_ass( char scope,
+        std::vector<diag_value> const &/* params */, diag_kwargs const &kwargs )
+{
+    diag_value school( std::string{} );
+    if( kwargs.count( "school" ) ) {
+        school = *kwargs.at( "school" );
+    }
+    diag_value spell( std::string{} );
+    if( kwargs.count( "spell" ) != 0 ) {
+        spell = *kwargs.at( "spell" );
+    }
+    return[beta = is_beta( scope ), school, spell]( dialogue const & d, double val ) {
+        const Character *ch = d.actor( beta )->get_character();
+        if( ch ) {
+            std::string school_str = school.str( d );
+            std::string spell_str = spell.str( d );
+            if( !school_str.empty() ) {
+                trait_id scid( school_str );
+                std::map<trait_id, double>::iterator it = ch->magic->caster_level_adjustment_by_school.find( scid );
+                if( it != ch->magic->caster_level_adjustment_by_school.end() ) {
+                    it->second = val;
+                } else {
+                    ch->magic->caster_level_adjustment_by_school.insert( { scid, val } );
+                }
+            } else if( !spell_str.empty() ) {
+                spell_id spid( spell_str );
+                std::map<spell_id, double>::iterator it = ch->magic->caster_level_adjustment_by_spell.find( spid );
+                if( it != ch->magic->caster_level_adjustment_by_spell.end() ) {
+                    it->second = val;
+                } else {
+                    ch->magic->caster_level_adjustment_by_spell.insert( { spid, val } );
+                }
+            } else {
+                ch->magic->caster_level_adjustment = val;
+            }
+            return val;
+        }
+        return 0.0;
+    };
+}
+
+std::function<double( dialogue & )> proficiency_eval( char scope,
+        std::vector<diag_value> const &params, diag_kwargs const &/* kwargs */ )
+{
+    diag_value format_value( std::string( "time_spent" ) );
+    if( params.empty() ) {
+        throw std::invalid_argument( string_format( "Not enough arguments for function %s()",
+                                     "proficiency" ) );
+    } else if( params.size() > 1 ) {
+        format_value = params[1];
+    }
```

Comment:
Not sure I like this usage of variadic functions. First issue is that you're duplicating this check a lot and the error message should be the parser's job. Second issue is that you're silently discarding the extra arguments beyond the second one.

Why not use a kwarg for `format_value`? I think it'd make more sense for JSON-only contributors too.

Reply:
I will replace it with kwarg